tests/LanguageServerNameImport/Service/NameImportServiceTest.php

```diff
@@ -0,0 +1,271 @@
+use Phpactor\CodeTransform\Domain\Refactor\ImportClass\NameAlreadyImportedException;
+use Phpactor\CodeTransform\Domain\Refactor\ImportClass\NameImport as ImportClassNameImport;
+use Phpactor\CodeTransform\Domain\Refactor\ImportName as RefactorImportName;
+use Phpactor\CodeTransform\Domain\SourceCode;
+use Phpactor\Extension\LanguageServerBridge\Converter\TextEditConverter;
+use Phpactor\Extension\LanguageServerNameImport\Service\NameImport;
+use Phpactor\LanguageServer\Core\Workspace\Workspace;
+use Phpactor\LanguageServerProtocol\TextDocumentItem;
+use Phpactor\LanguageServerProtocol\TextEdit as LspTextEdit;
+use Phpactor\TestUtils\PHPUnit\TestCase;
+use Phpactor\TextDocument\ByteOffset;
+use Phpactor\TextDocument\TextDocumentUri;
+use Phpactor\TextDocument\TextEdit;
+use Phpactor\TextDocument\TextEdits;
+use Prophecy\Prophecy\ObjectProphecy;
+use RuntimeException;
+
+class NameImportServiceTest extends TestCase
+{
+    const EXAMPLE_CONTENT = 'hello this is some text';
+    const EXAMPLE_PATH = '/foobar.php';
+    const EXAMPLE_OFFSET = 12;
+    const EXAMPLE_PATH_URI = 'file:///foobar.php';
+
+    /**
+     * @var ObjectProphecy<RefactorImportName>
+     */
+    private $importNameProphecy;
+
+    /**
+     * @var ObjectProphecy<Workspace>
+     */
+    private $workspaceProphecy;
+
+    /**
+     * @var ObjectProphecy<TextDocumentItem>
+     */
+    private $documentProphecy;
+
+    /**
+     * @var array<LspTextEdit>
+     */
+    private $lspTextEdits;
+
+    /**
+     * @var TextEdits
+     */
+    private $textEdits;
+
+    /**
+     * @var SourceCode
+     */
+    private $sourceCode;
+
+    /**
+     * @var ByteOffset
+     */
+    private $byteOffset;
+
+    /**
+     * @var ObjectProphecy<AliasAlreadyUsedException>
+     */
+    private $aliasAlreadyUsedExceptionProphecy;
+
+    /**
+     * @var ObjectProphecy<NameAlreadyImportedException>
+     */
+    private $nameAlreadyImportedExceptionProphecy;
+
+    /**
+     * @var NameImport
+     */
+    private $subject;
+
+    protected function setUp(): void
+    {
+        $this->documentProphecy = $this->prophesize(TextDocumentItem::class);
+        $this->documentProphecy->text = self::EXAMPLE_CONTENT;
+        $this->documentProphecy->uri = self::EXAMPLE_PATH_URI;
+
+        $this->workspaceProphecy = $this->prophesize(Workspace::class);
+        $this->workspaceProphecy->get(self::EXAMPLE_PATH_URI)
+            ->willReturn($this->documentProphecy->reveal());
+
+        $this->importNameProphecy = $this->prophesize(RefactorImportName::class);
+        $this->byteOffset = ByteOffset::fromInt(self::EXAMPLE_OFFSET);
+        $this->aliasAlreadyUsedExceptionProphecy = $this->prophesize(AliasAlreadyUsedException::class);
+        $this->nameAlreadyImportedExceptionProphecy = $this->prophesize(
+            NameAlreadyImportedException::class
+        );
+
+        $this->textEdits = TextEdits::one(
+            TextEdit::create(23, 6, 'huhuhu')
+        );
```

Comment:
My idea was to isolate the subject under test as good as possible.  
That is why I prophecised / mocked everything possible except `NameImport`.

Maybe just different opinions about test strategies. I am fine with either ways.

Reply:
yes - it's a concious choice to move away from doing this. Mocked dependencies do not act in the same way as the real objects, and some objects absolutely do not need to be mocked. Ideally we'd always test using real things to get the most value from a test, but practically we need to use fakes, stubs and mocks sometimes.

It might make sense to mock something with a complicated behavior and external dependencies (e.g. the NameImport class) but things that belong to the domain we're testing needn't be mocked. And some types of class should never be mocked (Value Objects, Exceptions, etc). We're testing behavior.